File: Java/src/Executor.java
import java.util.LinkedList;
import java.util.Queue;

public class Executor extends Thread{
    private Queue<Tarefa> tarefas;

    private TaskExecutor taskExecutor;
    private boolean possuiElementos;

    public Executor(TaskExecutor taskExecutor){
        this.taskExecutor = taskExecutor;
        tarefas = new LinkedList<Tarefa>();
        possuiElementos = true;
    }

    public void run(){
        Tarefa tarefa = null;
        while(!taskExecutor.Tarefas.isEmpty()){
            tarefa = taskExecutor.Tarefas.remove();
            //if(tarefa != null){
            despacharTarefa(tarefa);
            //}
        } 
        possuiElementos = false;

        if(tarefas.size() > 0) {
            acordarTrabalhadores();
        }
    }

    public void guardarResultado(Resultado resultado){
        taskExecutor.Resultados.add(resultado);
    }

	public synchronized void despacharTarefa(Tarefa tarefa) {
		tarefas.add(tarefa);
		notify();
	}

	public synchronized Tarefa pegarTarefa() {
		while (tarefas.size() == 0 && possuiElementos == true) {
			try {
				wait();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		
		return tarefas.poll();
	}

    public synchronized void acordarTrabalhadores(){
        notifyAll();
    }
}

File: GO/src/classes/Executor.go
package classes
import (
	"fmt"
	"time"
	"strconv"
)

type Executor struct {
	TaskExecutor    *TaskExecutor
	PossuiElementos bool
	CanalBufferizado chan<- *Tarefa
}

func CriarExecutor(taskExecutor *TaskExecutor, canalBufferizado chan<- *Tarefa) *Executor{
	return &Executor{
		TaskExecutor: taskExecutor,
		PossuiElementos: true,
		CanalBufferizado: canalBufferizado,
	}
}

func (e *Executor) Run() {
	inicioExecutor := time.Now();
	for len(e.TaskExecutor.Tarefas) > 0 {
		tarefa := e.TaskExecutor.Tarefas[0]
		e.TaskExecutor.Tarefas = e.TaskExecutor.Tarefas[1:]

		e.DespacharTarefa(tarefa)
	}
	fmt.Println("Executor = "+ strconv.FormatInt(time.Since(inicioExecutor).Milliseconds(), 10));
	close(e.CanalBufferizado)
}

func (e *Executor) DespacharTarefa(tarefa *Tarefa) {
	e.CanalBufferizado <- tarefa
}

func (e *Executor) GuardarResultado(resultado *Resultado) {
	e.TaskExecutor.Resultados = append(e.TaskExecutor.Resultados, resultado)
}
